### DISCUSSION TITLE
Sender in a box?

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/grblHAL#141] [General]
Openbuilds introduced "Interface"  [See it here.](https://openbuildspartstore.com/interface/)

I think it is basically a sender in a box but there aren't a lot of details so far. $180.

[Edit] [More details here](https://docs.openbuilds.com/doku.php?id=docs:interface:layout). It uses 5V serial I/O to connect to grbl. EMI anyone?

**Comment by frdfsnlght** [comment]
Isn't it basically a MPG/DRO interface? It's a nice idea, but I'd rather see something like that attached through a USB host interface on the controller. I think there was a discussion in one of the issues that was heading in that direction. That's the solution I'm aiming for. I'm going to end up with 2 or more CNC'ed machines and I'd like the ability to move the MPG/DRO thing between them as needed (not using the machines simultaneously).

**Reply by terjeio** [reply]
> Isn't it basically a MPG/DRO interface?

It is - and it seems to me it blocks any sender from connecting...

> It's a nice idea, but I'd rather see something like that attached through a USB host interface on the controller.

USB host interface would limit the controller selection but is possible. Note that some grblHAL drivers already has support for a secondary UART port and an input for switching between this and the primary port, I use this for the [MPG/DRO](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) I made for my lathe. Many drivers also has support for an I2C based keypad (via a [plugin](https://github.com/terjeio/grblHAL/tree/master/plugins/keypad)) that can be used for jogging etc.

**Reply by frdfsnlght** [reply]
Your MPG/DRO sounds a lot like what I was planning to build myself someday. It seems that if a controller offered a USB host interface, you could connect an MPR/DRO like yours to that with a simple USB serial port setup rather than needing an actual serial port.

**Reply by phil-barrett** [reply]
Terje's design uses i2c which limits the length of the connecting cable. For a DRO like device, it works quite well.  USB would be better for a traditional Pendant with a 2M/6ft cable.  Differential signaling of USB helps, too.

The T4.1 Pro board that I am working on has a host USB connection.  Though, currently no power managment.

**Reply by terjeio** [reply]
> Terje's design uses i2c 

It uses both serial and I2C - serial when the MPG is in control and acting as a sender. I2C is for the buttons only, they are usable even when the MPG is not in control. The I2C interface delivers data to the keypad plugin mentioned above.

**Reply by frdfsnlght** [reply]
Maybe it's abuse of some kind, but, at least in theory, could a USB connected MPG/DRO (even like yours, assuming just a USB serial port connection), use custom commands (M commands? I'm not sure of the GRBL terminology) to accomplish non-movement stuff? So, for instance, a button on the MPG to enable it could send a request to the controller and then the controller, assuming the correct state, could send the MPG:1 thing you use? The GRBL jog commands could be sent from the MPG/DRO for jogging, of course, and maybe some more of those custom commands for buttons. And the buttons could be mapped to GRBL stuff or used as macros, maybe via configuration. Could all that be done in a plugin?

The appeals of a USB connection, to me, are as Phil stated, differential signaling for longer cables, a standard connector, and the possibility to even plug the device into the PC running your sender and using it there. I really like the idea of a device that can be moved around like that. With a well defined command set/protocol, it could become the new open-source standard for this type of thing.

As far as power is concerned, it would be nice of the USB host port could provide USB standard currents, but that obviously doesn't work well if the controller is being powered from just USB itself. At least in my current plans, I'll be powering the controller from an independent 5V supply, so the controller would just be passing some of the plentiful power down the host port. I don't expect the MPG/DRO would need more than standard USB power levels.

**Reply by terjeio** [reply]
> Could all that be done in a plugin?

Nearly so. In order to receive commands from a USB serial port connection a (USB Host) API must be defined and implemented in the driver(s) capable of handling that. This since a plugin should _not_ access any hardware directly - by definition it should be possible to use with any driver that has an API abstracting the underlying hardware/low level drivers, in this case providing a stream inteface. The API must be able to operate in two modes, one "simple" mode when not in full control as a sender, the other when in full control. "Simple"  mode could be sending single character commands similar to those implemented by the keypad plugin. I would use single character (top bit set) to switch MPG mode on/off - there is no need for defining a M-code for that.

EDIT: full duplex RS-485 or RS-422 might be a better alternative to a USB Host based serial connection, most processors/cards supported by grblHAL could then be used with the addition of a (cheap) converter.

**Comment by phil-barrett** [comment]
They claim it passes through any Sender plugged into the "Interface" USB port. (terrible name, the Interface has an interface that you interface to.) I'm not sure if there is an official distinction between a sender and a pendant/mpg but the box can initiate a GCode stream to the Grbl box.  

I've whipped up a design for a little +5 to 3.3V serial daughter card for the Teensy BOB in case there is any interest.  It looks like a 2.54 mm pitch "duppont" style header on the Interface but the picture is terrible.  Right now, grblHAL or Teensy 4.1 needs to be built to use Serial. It would be interesting to have a switch to be able to do it at run time.  I don't suppose autosensing is easy.
![here's their picture](https://docs.openbuilds.com/lib/exe/fetch.php?cache=&w=454&h=700&tok=e5ad8b&media=docs:interface:bb_wiring.png)

I think we can claim Interface support once tested.  I suspect, however, the Interface interface to the Blackbox is a bit rough when you start connecting to other Grbl boards.

**Reply by terjeio** [reply]
>  I don't suppose autosensing is easy.

Monitor current draw on the 5V line?

**Comment by phil-barrett** [comment]
Actually, just +5V presence should be sufficient.  The header on T41 BOB would need an additional pin to let grblHAL know, I think.

**Comment by phil-barrett** [comment]
more info. It takes it's +5V power from the serial connection so any Grbl board will need to be able to supply the [yet to be documented] power draw of the Interface. I'm guessing it is in the 100 mA range. It also means that the Grbl board will need to be powered separately.

Also, if you are using a computer as well, you will need to have 2 cables - USB from the Computer and serial to the Grbl Controller.  I don't think that configuration will make hand held operations easy. It also looks like they designed it to be bolted onto the Blackbox for a stationary DRO/MPG.

**Reply by terjeio** [reply]
> They should have used Bluetooth for the connection? And a rechargeable battery for power?

BTW I got a [Nordic Semi M4F dev board today](https://www.nordicsemi.com/Software-and-Tools/Development-Kits/nRF52840-DK) - I am going to make a driver for that at some point, then the latest [BBC micro:bit](https://tech.microbit.org/latest-revision/announcement/) when I can get hold of one.

**Comment by Klippert** [comment]
Does anybody have a working implementation of a pendant with GRBL-Hal? i.e. to be able to jog the machine around, probing etc? 

I'm looking for a good place to start and if someone has something working already, it'd be great to hear about. Seems this sender in a box is not quite there yet, so wondered if any other solutions exist?


Thanks in advance
